src/lib/Type.hs

```diff
@@ -309,20 +309,27 @@ withAllowedEff eff m = updateAllowedEff (const eff) m
 
 -- === primitive ops and constructors ===
 
-typeCheckTyCon :: TC -> TypeM ()
+typeCheckTyCon :: TC -> TypeM Type
 typeCheckTyCon tc = case tc of
-  BaseType _       -> return ()
-  ArrayType t      -> t|:TyKind
-  IntRange a b     -> a|:IntTy >> b|:IntTy
+  BaseType _       -> return TyKind
+  ArrayType t      -> t|:TyKind >> return TyKind
+  IntRange a b     -> a|:IntTy >> b|:IntTy >> return TyKind
   IndexRange t a b -> t|:TyKind >> mapM_ (|:t) a >> mapM_ (|:t) b
-  SumType  l r     -> l|:TyKind >> r|:TyKind
-  PairType a b     -> a|:TyKind >> b|:TyKind
-  UnitType         -> return ()
-  RefType r a      -> r|:TyKind >> a|:TyKind
-  TypeKind         -> return ()
-  EffectRowKind    -> return ()
+                          >> return TyKind
+  SumType  l r     -> l|:TyKind >> r|:TyKind >> return TyKind
+  PairType a b     -> a|:TyKind >> b|:TyKind >> return TyKind
+  UnitType         -> return TyKind
+  RefType r a      -> r|:TyKind >> a|:TyKind >> return TyKind
+  TypeKind         -> return TyKind
+  EffectRowKind    -> return TyKind
   JArrayType _ _   -> undefined
-  NewtypeApp _ _ -> return () -- TODO!
+  NewtypeApp _ wrapTy xs -> foldM checkApp wrapTy xs where
+    checkApp :: Atom -> Atom -> TypeM Type
+    checkApp (Pi piTy) x = do
+      x |: absArgType piTy
+      -- Newtype arrows should be pure.
+      let (PlainArrow Pure, resultTy) = applyAbs piTy x
```

Comment:
It's better to perform checks like this one through code like
```hs
(PlainArrow Pure, resultTy) <- return $ applyAbs piTy x
```
because that will use the `MonadFail` instance of `TypeM` to print a nice error message, instead of trapping the program due to the pattern match failure in pure code.

Reply:
Oh, is that why that shows up everywhere? I didn't know that about Haskell!